nucypher/network/retrieval.py

```diff
@@ -56,8 +58,11 @@ class RetrievalPlan:
     def __init__(self, treasure_map: TreasureMap, retrieval_kits: Sequence[RetrievalKit]):
 
         # Record the retrieval kits order
-        self._capsules, _conditions = tuple(zip(*((rk.capsule, rk.conditions) for rk in retrieval_kits)))
-        self._conditions = list(ConditionLingo.from_bytes(l) if l else None for l in _conditions)
+        self._capsules, rust_conditions = tuple(zip(*((rk.capsule, rk.conditions) for rk in retrieval_kits)))
+
+        # Transform Conditions -> ConditionsLingo
+        json_conditions = (json.loads(str(c)) for c in rust_conditions)
+        self._conditions = list(ConditionLingo.from_list(lingo) if lingo else None for lingo in json_conditions)
```

Comment:
Is there a way to avoid this clear rust intrusion into python code - feels like a bit of a leak? I'm a little hazy with the intricacies of the rust layer/python bindings.

Something that could possibly be done with the python bindings, OR a helper class method in ConditionLingo for the conversion, OR utility function somewhere, OR ... (?)

Similar block occurs in server.py.

Reply:
I completely agree, in fact there's a TODO below that reads *"As this pattern swells further, it makes sense to do this in a purpose-built facility."*

This is certainly an abstraction leak and these lines are purely mechanical to mitigate the overlapping types in the rust and python layers (see https://github.com/nucypher/nucypher/issues/2968).

Let's have more discussion about this.